google/resource_cloudiot_registry.go

```diff
@@ -0,0 +1,381 @@
+	config := meta.(*Config)
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+	region, err := getRegion(d, config)
+	if err != nil {
+		return err
+	}
+	parent := fmt.Sprintf("projects/%s/locations/%s", project, region)
+
+	deviceRegistry := createDeviceRegistry(d)
+	deviceRegistry.Id = d.Get("name").(string)
+
+	call := config.clientCloudIoT.Projects.Locations.Registries.Create(parent, deviceRegistry)
+	res, err := call.Do()
+	if err != nil {
+		return err
+	}
+	d.SetId(res.Name)
+	return resourceCloudIoTRegistryRead(d, meta)
+}
+
+func resourceCloudIoTRegistryUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	updateMask := make([]string, 0, 5)
+	hasChanged := false
+	deviceRegistry := &cloudiot.DeviceRegistry{}
+
+	d.Partial(true)
+
+	if d.HasChange("event_notification_config") {
+		hasChanged = true
+		updateMask = append(updateMask, "event_notification_config")
+		if v, ok := d.GetOk("event_notification_config"); ok {
+			deviceRegistry.EventNotificationConfigs = make([]*cloudiot.EventNotificationConfig, 1, 1)
+			deviceRegistry.EventNotificationConfigs[0] = buildEventNotificationConfig(v.(map[string]interface{}))
+		}
+	}
+	if d.HasChange("state_notification_config") {
+		hasChanged = true
+		updateMask = append(updateMask, "state_notification_config")
+		if v, ok := d.GetOk("state_notification_config"); ok {
+			deviceRegistry.StateNotificationConfig = buildStateNotificationConfig(v.(map[string]interface{}))
+		}
+	}
+	if d.HasChange("mqtt_config") {
+		hasChanged = true
+		updateMask = append(updateMask, "mqtt_config")
+		if v, ok := d.GetOk("mqtt_config"); ok {
+			deviceRegistry.MqttConfig = buildMqttConfig(v.(map[string]interface{}))
+		}
+	}
+	if d.HasChange("http_config") {
+		hasChanged = true
+		updateMask = append(updateMask, "http_config")
+		if v, ok := d.GetOk("http_config"); ok {
+			deviceRegistry.HttpConfig = buildHttpConfig(v.(map[string]interface{}))
+		}
+	}
+	if d.HasChange("credentials") {
+		hasChanged = true
+		updateMask = append(updateMask, "credentials")
+		if v, ok := d.GetOk("credentials"); ok {
+			deviceRegistry.Credentials = expandCredentials(v.([]interface{}))
+		}
+	}
+	if hasChanged {
+		_, err := config.clientCloudIoT.Projects.Locations.Registries.Patch(d.Id(),
+			deviceRegistry).UpdateMask(strings.Join(updateMask, ",")).Do()
+		if err != nil {
+			return fmt.Errorf("Error updating registry %s: %s", d.Get("name").(string), err)
+		}
+		for _, updateMaskItem := range updateMask {
+			d.SetPartial(updateMaskItem)
+		}
+	}
+	d.Partial(false)
+	return resourceCloudIoTRegistryRead(d, meta)
+}
+
+func resourceCloudIoTRegistryRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	name := d.Id()
+	res, err := config.clientCloudIoT.Projects.Locations.Registries.Get(name).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Registry %q", name))
+	}
+
+	d.Set("name", res.Id)
```

Comment:
A comment here which explains that you're speculatively setting all attributes to nil and updating the ones you find later might help.

Reply:
After some testing I saw that all the nil checks are not necessary. Hope my changes now makes it more readable.
Just to make sure we have the same thing in mind when we read the code:
This is
```
resource "google_cloudiot_registry" "hello" {
  name = "hello"
}
```
comparable with
```
{
 "id": "hello",
 "name": "projects/<...>/locations/<...>/registries/hello",
 "mqttConfig": {
  "mqttEnabledState": "MQTT_ENABLED"
 },
 "stateNotificationConfig": {
 },
 "httpConfig": {
  "httpEnabledState": "HTTP_ENABLED"
 }
}
```
 because the code ignores default state, mqtt, http config: